### PR TITLE
[docs] Add documentation for reference counting and 'ray memory'

### PR DESCRIPTION
**Pull request by edoakes**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: doc/source/memory-management.rst [hunk 1]
```diff
@@ -1,10 +1,172 @@
...
+In this example, we create references to two objects: one that is ``ray.put()`` in the object store and another that's the return value from ``f.remote()``.
+
+-----------------------------------------------------------------------------------------------------
+Object ID                                Reference Type       Object Size   Reference Creation Site
+=====================================================================================================
+; driver pid=18867
+ffffffffffffffffffffffff0100008801000000  LOCAL_REFERENCE               77   (put object) ../test.py:<module>:9
+45b95b1c8bd3a9c4ffffffff010000c801000000  LOCAL_REFERENCE                ?   (task call) ../test.py:<module>:10
+-----------------------------------------------------------------------------------------------------
+
```
Comment (richardliaw) [inline doc/source/memory-management.rst]: Getting the following errors - 

you should probably put this in a code-block.

```
/Users/rliaw/Research/riselab/ray/doc/source/memory-management.rst:33: WARNING: Title overline & underline mismatch.

-----------------------------------------------------------------------------------------------------
Object ID                                Reference Type       Object Size   Reference Creation Site
=====================================================================================================
/Users/rliaw/Research/riselab/ray/doc/source/memory-management.rst:62: WARNING: Title overline & underline mismatch.

-----------------------------------------------------------------------------------------------------
Object ID                                Reference Type       Object Size   Reference Creation Site
=====================================================================================================
/Users/rliaw/Research/riselab/ray/doc/source/memory-management.rst:86: WARNING: Title overline & underline mismatch.

-----------------------------------------------------------------------------------------------------
Object ID                                Reference Type       Object Size   Reference Creation Site
=====================================================================================================
/Users/rliaw/Research/riselab/ray/doc/source/memory-management.rst:112: WARNING: Title overline & underline mismatch.

-----------------------------------------------------------------------------------------------------
Object ID                                Reference Type       Object Size   Reference Creation Site
=====================================================================================================
/Users/rliaw/Research/riselab/ray/doc/source/memory-management.rst:133: WARNING: Title overline & underline mismatch.

-----------------------------------------------------------------------------------------------------
Object ID                                Reference Type       Object Size   Reference Creation Site
=====================================================================================================

**Review by ericl — APPROVED** [review]
Very nice.

**Review by richardliaw — CHANGES_REQUESTED** [review]
Can you fix rendering first?

File: doc/source/memory-management.rst [hunk 2]
```diff
@@ -1,10 +1,172 @@
...
+**I'm running Ray inside IPython or a Jupyter Notebook and there are ObjectID references causing problems even though I'm not storing them anywhere.**
+
+IPython stores the output of every cell in a local Python variable indefinitely. This causes Ray to pin the objects even though your application may not actually be using them. You can try `Enabling LRU Fallback`_ to solve this, which will cause unused objects referenced by IPython to be LRU evicted when the object store is full instead of erroring.
+
+**My application used to run on previous versions of Ray but now I'm getting ObjectStoreFullError.**
+
+In previous versions of Ray, there was no reference counting and instead objects in the object store were LRU evicted once the object store ran out of space. Some applications (e.g., applications that keep references to all objects ever created) may have worked with LRU eviction but do not with reference counting. In this case, you can either modify your application to remove ``ObjectID`` references or try `Enabling LRU Fallback`_ to revert to the old behavior.
```
[the remark this answers is not in the document]
Comment (edoakes) [reply doc/source/memory-management.rst]: Good idea!

**Review by richardliaw — APPROVED** [review]
Pushed a lint fix - looks great!

Suspect the Jupyter notebook use case is going to be a big problem.

**Pull request merged**